semantic-core/src/Analysis/ScopeGraph.hs

```diff
@@ -1,15 +1,130 @@
+{-# LANGUAGE FlexibleContexts, OverloadedStrings, RankNTypes, RecordWildCards, TypeApplications, TypeOperators #-}
 module Analysis.ScopeGraph
-( ScopeGraph
-, Entry(..)
+( ScopeGraph(..)
+, Decl(..)
+, scopeGraph
+, scopeGraphAnalysis
 ) where
 
+import           Analysis.Eval
+import           Analysis.FlowInsensitive
+import           Control.Applicative (Alternative (..))
+import           Control.Effect.Carrier
+import           Control.Effect.Fail
+import           Control.Effect.Fresh
+import           Control.Effect.Reader
+import           Control.Effect.State
+import           Control.Monad ((>=>))
+import           Data.File
+import           Data.Foldable (fold)
+import           Data.Function (fix)
+import           Data.List.NonEmpty
 import           Data.Loc
 import qualified Data.Map as Map
+import           Data.Name
+import           Data.Proxy
 import qualified Data.Set as Set
+import           Data.Text (Text)
+import           Data.Traversable (for)
+import           Prelude hiding (fail)
 
-data Entry = Entry
-  { entrySymbol :: String -- FIXME: Text
-  , entryLoc    :: Loc
+data Decl = Decl
+  { declSymbol :: Text
+  , declLoc    :: Loc
   }
+  deriving (Eq, Ord, Show)
 
-type ScopeGraph = Map.Map Entry (Set.Set Entry)
+newtype Ref = Ref Loc
+  deriving (Eq, Ord, Show)
+
+newtype ScopeGraph = ScopeGraph { unScopeGraph :: Map.Map Decl (Set.Set Ref) }
```

Comment:
This model of scope graphs has no recursive structure; references map directly to declarations, and not e.g. through other declarations in a path.

This makes this scope graph strictly less informative than those we’re computing in `semantic` proper, tho it’s hard to say for sure if that makes it strictly less useful for our purposes.

Reply:
Unlike `ImportGraph`, our `ScopeGraph` abstract domain doesn’t require a semi-concrete representation of values alongside it, because we don’t need to treat any of the primitives specially: `ImportGraph` needs strings and closures modelled explicitly so that it can notice loads issued within methods like Ruby’s `require`, but `ScopeGraph` doesn’t require any special treatment of primitives (other than to construct a scope graph).